src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -449,6 +449,7 @@ private unsafe void CloseRequestQueueHandle()
             if ((_requestQueueHandle != null) && (!_requestQueueHandle.IsInvalid))
             {
                 if (NetEventSource.IsEnabled) NetEventSource.Info($"Dispose ThreadPoolBoundHandle: {_requestQueueBoundHandle}");
+                Interop.Kernel32.CancelIoEx(_requestQueueHandle, null);
```

Comment:
Is there any return value from calling this API? It might be useful to log it if NetEventSource.IsEnabled is true.

Reply:
We already have [code calling CancelIoEx](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpResponseStream.Windows.cs#L390) which explicitly says we don't need to check the returned value. However, if you still think this can be useful, I can add the logging.

Reply:
Is it worth adding a comment on why you're calling cancel? E.g. to unblock sync operations.

There is a small race here. You could be starting a call to GetContext on anther thread between the cancel and the Dispose. Could you take advantage of the fact that the ref counting keeps the handle open and call Dispose before Cancel? That risks an ObjectDisposedException from CancelIoEx, but that can be caught. Is the ODE still thrown in the affected scenario?

Reply:
> Could you take advantage of the fact that the ref counting keeps the handle open and call Dispose before Cancel? That risks an ObjectDisposedException from CancelIoEx, but that can be caught.

Yes, that is the right thing to do, otherwise as you say there's a race condition where synchronous calls could still hang.

Reply:
Agree, the current code is vulnerable to race condition. I will fix it as proposed above. One question, though. I noticed that _requestQueueHandle is **not volatile**, so I'm concerned whether the handle read operations can be reordered in the following case?

```C#
_requestQueueHandle.Dispose();
// !!!GetContext is called at this point and _requestQueueHandle is recreated
try
{
  Interop.Kernel32.CancelIoEx(_requestQueueHandle, null);
}
catch (ObjectDisposedException)
{
  // ignore the exception since it only means that the queue handle has been successfully disposed
}

Reply:
> so I'm concerned whether the handle read operations can be reordered in the following case?

I'm not following.  What do you mean by _requestQueueHandle being recreated?  Are we setting _requestQueueHandle to null somewhere?  If you're referring to the underlying handle, that's not affected by whether _requestQueueHandle is volatile or not.

Reply:
I meant the scenario Start -> GetContext -> Stop -> Start (while we are between Dispose and CloseEx in Stop). However, I see now that I made two mistakes. First, a new request handle is created inside Start, not GetContext. Second, Start and Stop are synchronized via ```lock```. Thus, requestQueueHandle **cannot** be recreated while we are inside Stop.

Reply:
Pushed version with CloseIoEx being called after Dispose.